### ISSUE TITLE
Query with "failOnClusterChange" breaks command queue

### ISSUE THREAD

**Issue by Aconradty** [aerospike/aerospike-client-nodejs#389]
Queries with "failOnClusterChange" and secondary index filter are instantly timing out once they are getting queued.
The test passes correctly if  'maxCommandsInProcess' setting is high enough or if no secondary index is used.

Test to reproduce:
```
it('queue with failOnClusterChange', async (done) => {
  Aerospike.setupGlobalCommandQueue({ maxCommandsInProcess: 5, maxCommandsInQueue: 50 })
  client = await Aerospike.connect({
    hosts: '127.0.0.1:3000',
    policies: {
      query: new Aerospike.QueryPolicy({ totalTimeout: 10000, failOnClusterChange: true })
    }
  })

  try {
    await client.createIntegerIndex({
      ns: 'test',
      set: 'test',
      bin: 'i',
      index: 'test'
    })
  } catch (e) {
    //
  }

  const puts = Array.from({ length: 5 }, (_, i) => client.put(new Aerospike.Key('ssd-store', 'test', i), { i }))

  await Promise.all(puts)

  const queries = Array.from({ length: 5 }, async (_, i) => {
    const query = client.query('ssd-store', 'test')

    // does not work with sencondary index
    query.where(filter.equal('i', i))
    // works with predexp
    // query.where([predexp.integerBin('i'), predexp.integerValue(5), predexp.integerGreaterEq()])
    await query.results()
  })

  await Promise.all(queries)
  client.close()
  done()
})
```

Result: ` AerospikeError: Delay queue timeout`

**Comment by fschaeffler** [comment]
Also experiencing the same issue. :+1: for the test case!

**Comment by jhecking** [comment]
Thanks for reporting this issue. I'm able to replicate it, though I had to make some minor modifications to your test:

```javascript
const Aerospike = require('../lib/aerospike')
const { predexp, filter } = Aerospike

let client
before(async () => {
  Aerospike.setupGlobalCommandQueue({ maxCommandsInProcess: 5, maxCommandsInQueue: 50 })
  Aerospike.setDefaultLogging({ level: Aerospike.log.WARN, file: process.stderr.fd })
  client = await Aerospike.connect({
    policies: {
      query: new Aerospike.QueryPolicy({ totalTimeout: 10000, failOnClusterChange: true })
    }
  })
})

after(() => {
  client.close()
})

it('queue with failOnClusterChange', async () => {
  try {
    await client.createIntegerIndex({
      ns: 'test',
      set: 'test',
      bin: 'i',
      index: 'test'
    })
  } catch (e) {
    // console.error(e)
  }

  const puts = Array.from({ length: 5 }, (_, i) => client.put(new Aerospike.Key('test', 'test', i), { i }))
  await Promise.all(puts)

  let results = Array.from({ length: 5 }, async (_, i) => {
    const query = client.query('test', 'test')

    // does not work with sencondary index
    query.where(filter.equal('i', i))
    // works with predexp
    // query.where([predexp.integerBin('i'), predexp.integerValue(i), predexp.integerGreaterEq()])
    return query.results()
  })
  results = await Promise.all(results)
  console.log(results.map((r) => r.pop()))
})
```

**Comment by fschaeffler** [comment]
@jhecking thank you for addressing this! As this is quite a serious issue for us, do you have a rough timeline for when a fix might be available?

**Comment by jhecking** [comment]
Not yet, as I do not yet have full clarity what the issue is exactly.

What I know is, that depending on the server version, setting `failOnClusterChange` to `true` will cause the client to send an extra info command to the cluster, before it will start to process the query. Those info commands are using the same command queue, so if you already have 5 pending query commands in the queue and `maxCommandsInProcess` is set to 5 as well, that might cause a deadlock. But I'm as of yet unclear what role the secondary index filter plays in this. I've reached out to some other team members for some additional insights.

Can I check with you what server version you are running?

**Comment by fschaeffler** [comment]
We're IMO running server and client at a decently updated state. 

Client from `package.json`: `"aerospike": "3.16.1"`
Server via `asinfo -v build`: `5.2.0.5`

via `aql`:
```
aql> GET FAIL_ON_CLUSTER_CHANGE
FAIL_ON_CLUSTER_CHANGE = true
```

**Comment by jhecking** [comment]
Quick update: The secondary index filter is relevant because without it the query is actually processed as a scan, and for scan requests with `failOnClusterChange = true` the client does not send the extra info command.

The initial fix will have to be done in the Aerospike C client. And given that this week is Thanks Giving in the US that will happen some time next week at the earliest. And even that is still TBC because it's not yet clear what the best way to resolve this issue is. Once a new C client version with the fix has been released, I will still have to update the C client version used within the Node.js client and do a new release for that as well.

**Comment by fschaeffler** [comment]
@jhecking thank you again for addressing the issue! If I can help with the Node.js part, just let me know.

When it comes to C, I'm, unfortunately - or luckily depending on the point of view :speak_no_evil: - not proficient enough to tackle the issue myself.

**Comment by jhecking** [comment]
A new C client version 4.6.19 is ready, which includes a fix for this issue. I'm planning to release Node.js client version 3.16.2 tomorrow.

**Comment by jhecking** [comment]
I've just released v3.16.2 to [npmjs.com](https://www.npmjs.com/package/aerospike).

**Comment by fschaeffler** [comment]
This awesome! Thank you so much for your quick help here.
